File: dongwook/week7/두 큐 합 같게 만들기/두큐합같게만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = -2;

        long sum1 = 0;
        long sum2 = 0;

        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();

        for (int i = 0; i < queue1.length; i++) {
            sum1 += queue1[i];
            sum2 += queue2[i];
            q1.add(queue1[i]);
            q2.add(queue2[i]);
        }

        long target = (sum1 + sum2) / 2;

        while (sum1 != sum2) {
            if (answer >= queue1.length * 3) // 큐 최대 3번까지 순회하여 합을 맞춰볼 수 있음
                return -1;

            if (sum1 > sum2) {
                sum1 -= q1.peek();
                q2.add(q1.peek());
                sum2 += q1.poll();
            } else {
                sum2 -= q2.peek();
                q1.add(q2.peek());
                sum1 += q2.poll();
            }
            answer++;
        }
        return answer;
    }
}

File: seongmin/week7/두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    queue<int> q1, q2;
    long long sum1 = 0, sum2 = 0, total = 0;

    for (int num : queue1) {
        q1.push(num);
        sum1 += num;
    }
    for (int num : queue2) {
        q2.push(num);
        sum2 += num;
    }

    total = sum1 + sum2;
    if (total % 2 != 0) return -1;

    long long target = total / 2;
    int maxOperations = 2 * (queue1.size() + queue2.size());
    int operations = 0;

    while (sum1 != target) {
        if (operations > maxOperations) return -1;

        if (sum1 > target) {
            int front = q1.front();
            q1.pop();
            q2.push(front);
            sum1 -= front;
            sum2 += front;
        }
        else {
            int front = q2.front();
            q2.pop();
            q1.push(front);
            sum2 -= front;
            sum1 += front;
        }
        operations++;
    }

    return operations;
}
